extraction/src/main/scala/org/broadinstitute/monster/encode/extraction/ExtractionPipelineBuilder.scala

```diff
@@ -24,52 +21,86 @@ import scala.concurrent.Future
+
+  private val lookupFn =
+    new ScalaAsyncLookupDoFn[LookupParams, Msg, EncodeClient](MaxConcurrentRequests) {
+
+      override def asyncLookup(
+        client: EncodeClient,
+        params: LookupParams
+      ): Future[Msg] = {
+        val (encodeEntity, filters) = params
+        client.get(encodeEntity, filters)
+      }
+      override protected def newClient(): EncodeClient = getClient()
+    }
+
+  /**
+    * Map a  batch of search parameters into MessagePack entities from
+    * ENCODE matching those parameters.
+    *
+    * @param encodeEntity the type of ENCODE entity the stage should query
+    * @param filterBatches batches of key=value filters to include in API queries
+    */
+  def getEntities(
+    encodeEntity: EncodeEntity,
+    filterBatches: SCollection[List[(String, String)]]
+  ): SCollection[Msg] =
+    filterBatches
+      .withName("Construct full query parameters")
+      .map(encodeEntity -> _)
+      .withName("Query ENCODE API")
+      .applyKvTransform(ParDo.of(lookupFn))
+      .withName("Extract result graph")
+      .flatMap(_.getValue.fold(throw _, _.read[Array[Msg]]("@graph")))
+
   override def buildPipeline(ctx: ScioContext, args: Args): Unit = {
     /**
       * Generic helper method for extracting linked entities and saving them.
       *
-      * @param entityToExtract the entity which should be extracted
-      * @param matchingField name of the field to use as an identifier for the entityToExtract
-      *                      (should match the id that's used in the linkedField)
-      * @param linkingEntities the parent entity (which references a set of encodeEntities)
-      *                        from which ID extraction will be based
-      * @param linkedField name of the entityToExtract field which references the linkingEntities (default is @id)
-      * @return the extracted linked entities
+      * @param sourceEntityType type of objects contained in `sourceEntities`
+      * @param sourceField key within `sourceEntities` to extract for use as a query filter
+      * @param sourceEntities raw objects of type `sourceEntityType` containing data
+      *                       which can be used to pull `targetEntityType` objects
+      * @param targetEntityType type of objects to extract from the API
+      * @param targetField field in `targetEntityType` objects with values that match the
+      *                    `sourceField` values from `sourceEntityType` objects
       */
     def extractLinkedEntities(
-      entityToExtract: EncodeEntity,
-      matchingField: String,
-      linkingEntities: SCollection[Msg],
-      linkedField: String = "@id"
+      sourceEntityType: EncodeEntity,
+      sourceField: String,
+      sourceEntities: SCollection[Msg],
+      targetEntityType: EncodeEntity,
+      targetField: String
     ): SCollection[Msg] = {
-      linkingEntities.transform(
-        s"Extract ${matchingField} from ${entityToExtract.entryName} data"
```

Comment:
This description was very misleading: it made it sound like the thing we were extracting was already present, and we were pulling some field out of it. Reworded below,